### PR TITLE
New CompositeGate Class for Arbitrary Multi-Qubit Initialization

### PR DESCRIPTION
**Pull request by ismaila-at-za-ibm**
## Description
The main contribution is a new class Initializer that implements the (complex amplitude) initialization of a flexible collection of qubit registers (assuming the qubits are in the zero state).

 The class implements a recursive initialization algorithm including optimizations from "Synthesis of Quantum Logic Circuits" Shende, Bullock, Markov https://arxiv.org/abs/quant-ph/0406176v5

Additionally implements some extra optimizations: remove zero rotations and double cnots.

## Motivation and Context
The current SDK does not allow for arbitrary multi-qubit initialization.

## How Has This Been Tested?
Two python demo examples included (probably needs more testing).
No other code is affected because ONLY new functionality is added in the form of a new class that inherits from the CompositeGate class.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: qiskit/extensions/quantum_initialiser/initialiser.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+            # perform the required rotations to decouple the LSB qubit (so that it can be "factored" out, leaving a
+            # shorter amplitude vector to peel away)
+            self._attach(self._multiplex(RZGate, i, phis))
+            self._attach(self._multiplex(RYGate, i, thetas))
+
+    @staticmethod
+    def _log2(orig):
+        """Static internal method to calc the log (base 2) of the input (greater than 2). Return zero if the input is
+        not a positive power of 2. """
```
Comment (diego-plan9) [inline qiskit/extensions/quantum_initialiser/initialiser.py]: If this function is only used twice: once during `__init__` for checking that `param` is valid, and later in `multiplex()`, would it make sense to just check that it is a multiple of two (`param % 2 == 0`) during `__init__`, and then just rely on `numpy.log2`?

File: qiskit/extensions/quantum_initialiser/initialiser.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+    @staticmethod
+    def chop_num(numb):
+        """Set very small numbers (as defined by global variable EPS) to zero"""
+        return 0 if abs(numb) < EPS else numb
+
+
+# ###############################################################
+# Add needed functionality to other classes (it feels
+#  weird following the QISKit convention of adding functionality to other classes like this ;),
+#  TODO: multiple inheritance might be better?)
```
Comment (diego-plan9) [inline qiskit/extensions/quantum_initialiser/initialiser.py]: > it feels weird following the QISKit convention of adding functionality to other classes like this

It feels a bit unusual indeed, and comes at a discoverability/readability/maintainability cost that you seem to have experienced already, specially due to the high number of functions in the PRs - thanks for raising the issue! It seems a bit out of the scope of this PR to revise the convention, but I'll make a note so we can revise it eventally,

File: qiskit/extensions/quantum_initialiser/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from .initialiser import Initialiser
```
Comment (diego-plan9) [inline qiskit/extensions/quantum_initialiser/__init__.py]: Can you revise this import so it mimics the mechanism used in `qiskit/extensions/standard/__init__.py` (which will also most likely include adding a `import qiskit.extensions.quantum_initialiser` to the main `qiskit/__init__.py`?

The idea would be to make `QuantumCircuit.initialise()` (and the rest of the functions) available when importing qiskit, ie making the `from qiskit.extensions.quantum_initialiser import Initialiser` lines on the examples not needed.

**Comment by ismaila-at-za-ibm** [comment]
Reworked separately, resubmitting as a new pull request.

**Pull request closed without merging**